napari/utils/_injection/_processors.py

```diff
@@ -0,0 +1,230 @@
+
+from ... import layers, types, viewer
+from ...layers._source import layer_source
+from ...utils.misc import ensure_list_of_layer_data_tuple
+from ._accessors import _access_viewer
+
+T = TypeVar("T")
+_NULL = object()
+
+
+def _add_layer_data_tuples_to_viewer(
+    data: Any, return_type=None, viewer=None, source: Optional[dict] = None
+):
+    if viewer is None:
+        viewer = _access_viewer()
+    if viewer and data is not None:
+        data = data if isinstance(data, list) else [data]
+        for datum in ensure_list_of_layer_data_tuple(data):
+            # then try to update a viewer layer with the same name.
+            if len(datum) > 1 and (name := datum[1].get("name")):
+                with suppress(KeyError):
+                    layer = viewer.layers[name]
+                    layer.data = datum[0]
+                    for k, v in datum[1].items():
+                        setattr(layer, k, v)
+                    continue
+            with layer_source(**source) if source else nullcontext():
+                # otherwise create a new layer from the layer data
+                viewer._add_layer_from_data(*datum)
+
+
+def _add_layer_data_to_viewer(
+    data: Any,
+    return_type: Any,
+    viewer: Optional[viewer.Viewer] = None,
+    layer_name: Optional[str] = None,
+    source: Optional[dict] = None,
+):
+    """Show a result in the viewer.
+
+    Parameters
+    ----------
+    data : Any
+        The result of the function call. For this function, this should be
+        *just* the data part of the corresponding layer type.
+    return_type : Any
+        The return annotation that was used in the decorated function.
+    viewer : Optional[Viewer]
+        An optional viewer to use. Otherwise use current viewer.
+    layer_name : Optional[str]
+        An optional layer name to use. If a layer with this name exists, it will
+        be updated.
+    source : Optional[dict]
+        An optional layer source to use.
+
+    Examples
+    --------
+    This allows the user to do this, and add the result as a viewer Image.
+
+    >>> def make_layer() -> napari.types.ImageData:
+    ...     return np.random.rand(256, 256)
+
+    """
+    if data is not None and (viewer := viewer or _access_viewer()):
+        if layer_name:
+            with suppress(KeyError):
+                viewer.layers[layer_name].data = data
+                return
+        layer_type = return_type.__name__.replace("Data", "").lower()
+        with layer_source(**source) if source else nullcontext():
+            getattr(viewer, f'add_{layer_type}')(data=data, name=layer_name)
+
+
+def _add_layer_to_viewer(
+    layer: layers.Layer,
+    viewer: Optional[viewer.Viewer] = None,
+    source: Optional[dict] = None,
+):
+    if layer is not None and (viewer := viewer or _access_viewer()):
+        layer._source = layer.source.copy(update=source or {})
+        viewer.add_layer(layer)
+
+
+_FUTURES: Set[Future] = set()
```

Comment:
i believe it was necessary to prevent garbage collection before the future is done.  will add a comment

Reply:
Ah, I guess that's because normally you would create a Future with an executor (or similar), which probably does keep a reference. Explanatory comment sounds good.